compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -2863,6 +2863,8 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
         disjointnessBoundary(tp.effectiveBounds.hi)
       case tp: ErrorType =>
         defn.AnyType
+      case NoType => // ParamRef#superTypeNormalized can return NoType
```

Comment:
That seems to be a problem with `ParamRef#superTypeNormalized` rather than with `disjointnessBoundary`.

Reply:
What should `superTypeNormalized` do?

Reply:
It should return the declared type of the corresponding param, which is supposed to be its `underlying`. I don't see a scenario in which `superTypeNormalized` should ever return `NoType`, *except* possibly for `AnyKind` itself.

Reply:
Apparently, it can happen that it is not initialised:
https://github.com/lampepfl/dotty/blob/e994cf06eeee23e8a55a086907957e43140f2112/compiler/src/dotty/tools/dotc/core/Types.scala#L4690-L4695

Reply:
That sort of explains it, although it's still concerning that we're attempting to read it from `provablyDisjoint` when it hasn't been initialized yet. Looks like a time bomb waiting to explode later due to weird compilation order issues.

Reply:
Yes, and this is not the only case, I ran into a similar issue where normalisation relies on the `underlying` type, which is not always set as expected. For a `TypeRef`, it goes through the denotation info, of which the completion seemed a bit unpredictable to me (heavily influenced by small changes in where things happen to be reduced). 

Also (maybe) unrelated the normalization depends on the gadt ctx which doesn't seem to be reliably populated when using `NamedType`s.

Reply:
I guess I should've opted for AnyKind rather than Any, as the more conservative approach.

Reply:
I would still be less uncomfortable if we specifically handled `ParamRef` to deal with its `superTypeNormalized` returning `NoType`; instead of having this "catch-all" `NoType`. That would limit the blast radius of this workaround. We should also clearly comment that it is a workaround.

Reply:
Sure thing.